File: packages/python/elizaos/features/advanced_capabilities/personality/evaluators/character_evolution.py
"""Character evolution evaluator.

Analyzes conversations for character evolution opportunities. Runs after
conversations to identify patterns suggesting character growth, using a
two-tier approach: LLM-based trigger detection with regex fallback.
"""

from __future__ import annotations

import json
import logging
import time
from typing import TYPE_CHECKING, Any

from elizaos.types import ActionResult, Evaluator, MemoryType, ModelType
from elizaos.utils.xml import parse_key_value_xml

if TYPE_CHECKING:
    from elizaos.types import IAgentRuntime, Memory, State

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Pure helpers
# ---------------------------------------------------------------------------


def _is_record(value: Any) -> bool:
    return isinstance(value, dict)


def _parse_structured_record(response: str) -> dict[str, Any] | None:
    parsed = parse_key_value_xml(response)
    return parsed if _is_record(parsed) else None


def _normalize_boolean(value: Any) -> bool | None:
    if isinstance(value, bool):
        return value
    if isinstance(value, str):
        normalized = value.strip().lower()
        if normalized == "true":
            return True
        if normalized == "false":
            return False
    return None


def _normalize_number(value: Any) -> float | None:
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return float(value) if value == value else None  # NaN check
    if isinstance(value, str):
        trimmed = value.strip()
        if not trimmed:
            return None
        try:
            return float(trimmed)
        except ValueError:
            return None
    return None


def _normalize_string_list(value: Any) -> list[str] | None:
    if not isinstance(value, str):
        return None
    trimmed = value.strip()
    if not trimmed:
        return None
    delimited = [entry.strip() for entry in trimmed.split("||") if entry.strip()]
    return delimited if delimited else None


def _coerce_int(value: Any, default: int = 0) -> int:
    if isinstance(value, bool):
        return default
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return int(value)
    if isinstance(value, str):
        trimmed = value.strip()
        if trimmed:
            try:
                return int(trimmed)
            except ValueError:
                return default
    return default


def parse_trigger_analysis(raw: dict[str, Any]) -> dict[str, Any]:
    return {
        "hasEvolutionTrigger": _normalize_boolean(raw.get("hasEvolutionTrigger")) or False,
        "triggerType": raw.get("triggerType", "unknown")
        if isinstance(raw.get("triggerType"), str)
        else "unknown",
        "reasoning": raw.get("reasoning", "") if isinstance(raw.get("reasoning"), str) else "",
        "confidence": _normalize_number(raw.get("confidence")) or 0,
    }


def build_modifications(raw: dict[str, Any]) -> dict[str, Any]:
    mods: dict[str, Any] = {}

    name = raw.get("name")
    if isinstance(name, str) and name.strip():
        mods["name"] = name.strip()

    system = raw.get("system")
    if isinstance(system, str) and system.strip():
        mods["system"] = system.strip()

    bio = _normalize_string_list(raw.get("bio"))
    if bio:
        mods["bio"] = bio

    topics = _normalize_string_list(raw.get("topics"))
    if topics:
        mods["topics"] = topics

    style: dict[str, list[str]] = {}
    for key, field in [("all", "style_all"), ("chat", "style_chat"), ("post", "style_post")]:
        items = _normalize_string_list(raw.get(field))
        if items:
            style[key] = items
    if style:
        mods["style"] = style

    return mods


def parse_evolution_analysis(raw: dict[str, Any]) -> dict[str, Any]:
    return {
        "shouldModify": _normalize_boolean(raw.get("shouldModify")) or False,
        "confidence": _normalize_number(raw.get("confidence")) or 0,
        "gradualChange": _normalize_boolean(raw.get("gradualChange"))
        if raw.get("gradualChange") is not None
        else True,
        "reasoning": raw.get("reasoning", "") if isinstance(raw.get("reasoning"), str) else "",
        "modifications": build_modifications(raw),
    }


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------


async def _validate(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> bool:
    # Skip if too recent since last evaluation
    last_check = await runtime.get_cache("character-evolution:last-check")
    now = int(time.time() * 1000)
    cooldown_ms = 5 * 60 * 1000  # 5 minutes

    if last_check:
        try:
            if now - _coerce_int(last_check) < cooldown_ms:
                return False
        except (ValueError, TypeError):
            pass

    # Only evaluate if conversation has substantial content
    if state and hasattr(state, "data") and isinstance(state.data, dict):
        msg_count = state.data.get("messageCount", 0)
        if isinstance(msg_count, (int, float)) and msg_count < 3:
            return False

    # Get recent messages for trigger detection
    recent_messages = await runtime.get_memories(
        {
            "roomId": str(message.room_id),
            "count": 10,
            "unique": True,
            "tableName": "messages",
        }
    )

    # LLM-based trigger detection
    conversation_text = "\n".join(
        f"{'Agent' if m.entity_id == runtime.agent_id else 'User'}: {m.content.text}"
        for m in recent_messages
        if hasattr(m, "content") and hasattr(m.content, "text") and m.content.text
    )

    trigger_prompt = f"""Analyze this conversation for character evolution triggers:

CONVERSATION:
{conversation_text}

TRIGGER ANALYSIS - Check for:

1. CONVERSATION SUCCESS PATTERNS
   - User engagement (long responses, follow-up questions)
   - Positive sentiment from user

2. KNOWLEDGE GAP DISCOVERY
   - Agent uncertainty or "I don't know" responses
   - User providing corrections or new information

3. PERSONALITY EFFECTIVENESS
   - User preferences for communication style
   - Emotional intelligence opportunities

4. VALUE CREATION OPPORTUNITIES
   - User goals mentioned that agent could help with better

5. EXPLICIT FEEDBACK
   - Direct requests for personality changes
   - User feedback about agent behavior

TOON only. Return exactly one TOON document. No prose before or after it.

Example:
hasEvolutionTrigger: true
triggerType: explicit_feedback
reasoning: User explicitly asked for a personality change
confidence: 0.85"""

    has_triggers = False
    try:
        trigger_response = await runtime.use_model(
            ModelType.TEXT_SMALL,
            prompt=trigger_prompt,
        )
        raw = _parse_structured_record(str(trigger_response))
        if raw:
            trigger = parse_trigger_analysis(raw)
            has_triggers = trigger["hasEvolutionTrigger"] and trigger["confidence"] > 0.6

            if has_triggers:
                logger.info(
                    "Evolution trigger detected: type=%s, reasoning=%s, confidence=%s",
                    trigger["triggerType"],
                    trigger["reasoning"],
                    trigger["confidence"],
                )
    except Exception:
        # Fallback to basic pattern matching
        has_triggers = any(
            any(
                kw in (m.content.text or "").lower()
                for kw in [
                    "you should",
                    "change your",
                    "different way",
                    "personality",
                    "behavior",
                    "remember that",
                    "from now on",
                ]
            )
            for m in recent_messages
            if hasattr(m, "content") and hasattr(m.content, "text")
        )

    return has_triggers


async def _handler(
    runtime: IAgentRuntime,
    message: Memory,
    _state: State | None = None,
) -> ActionResult | None:
    try:
        await runtime.set_cache(
            "character-evolution:last-check",
            str(int(time.time() * 1000)),
        )

        # Get recent conversation context
        recent_messages = await runtime.get_memories(
            {
                "roomId": str(message.room_id),
                "count": 20,
                "unique": True,
                "tableName": "messages",
            }
        )

        conversation_text = "\n".join(
            f"{runtime.character.name if m.entity_id == runtime.agent_id else 'User'}: {m.content.text}"
            for m in recent_messages[-10:]
            if hasattr(m, "content") and hasattr(m.content, "text") and m.content.text
        )

        character = runtime.character
        character_summary = {
            "name": getattr(character, "name", ""),
            "system": getattr(character, "system", "No system prompt defined"),
            "bio": getattr(character, "bio", []),
            "currentTopics": getattr(character, "topics", []) or [],
            "messageExampleCount": len(getattr(character, "message_examples", []) or []),
        }

        bio_list = character_summary["bio"]
        if isinstance(bio_list, str):
            bio_list = [bio_list]
        elif not isinstance(bio_list, list):
            bio_list = []
        current_topics = character_summary["currentTopics"]
        topics_list = current_topics if isinstance(current_topics, list) else []

        evolution_prompt = f"""You are conducting a comprehensive analysis to determine if an AI agent should evolve its character definition based on measurable patterns and outcomes.

CURRENT CHARACTER STATE:
Name: {character_summary["name"]}
System: {character_summary["system"]}
Bio: {"; ".join(str(b) for b in bio_list)}
Topics: {", ".join(str(t) for t in topics_list)}
Message Examples: {character_summary["messageExampleCount"]}

CONVERSATION TO ANALYZE:
{conversation_text}

EVOLUTION DECISION FRAMEWORK:
- Only suggest modifications that address measurable gaps
- Prioritize changes that improve user experience
- Ensure gradual, incremental evolution
- Maintain core personality while optimizing effectiveness
- Consider safety and appropriateness of all changes

MODIFICATION PRIORITIES:
- name: ONLY if a truly fitting identity emerges organically (very rare)
- system: ONLY for fundamental behavioral misalignment (rare)
- bio: New traits that emerge from successful interactions
- topics: Domains where agent showed interest/competence
- style: Communication preferences that enhance effectiveness

TOON only. Return exactly one TOON document. No prose before or after it.
Use || to separate list items within a field.

Example:
shouldModify: true
confidence: 0.75
gradualChange: true
reasoning: User consistently asks about sustainability topics that are not in the current character definition.
bio: Passionate about environmental sustainability || Knowledgeable about renewable energy
topics: climate change || solar energy || sustainable living
style_chat: Use encouraging tone when discussing environmental topics"""

        response = await runtime.use_model(
            ModelType.TEXT_LARGE,
            prompt=evolution_prompt,
        )

        raw = _parse_structured_record(str(response))
        if not raw:
            logger.warning("Failed to parse character evolution analysis")
            return None

        evolution = parse_evolution_analysis(raw)

        # Only proceed if modification is recommended with sufficient confidence
        if not evolution["shouldModify"] or evolution["confidence"] < 0.7:
            return None

        if not evolution.get("gradualChange", True):
            logger.info("Skipping character evolution - change too dramatic")
            return None

        # Store evolution suggestion for potential application
        await runtime.create_memory(
            {
                "entityId": str(runtime.agent_id),
                "roomId": str(message.room_id),
                "content": {
                    "text": f"Character evolution suggested (confidence: {evolution['confidence']}): {evolution['reasoning']}",
                    "source": "character_evolution",
                },
                "metadata": {
                    "type": MemoryType.CUSTOM,
                    "evaluatorName": "character-evolution",
                    "timestamp": int(time.time() * 1000),
                    "confidence": evolution["confidence"],
                    "evolutionData": json.dumps(
                        {
                            "shouldModify": evolution["shouldModify"],
                            "gradualChange": evolution.get("gradualChange", True),
                            "modifications": evolution["modifications"],
                        }
                    ),
                },
            },
            "character_evolution",
        )

        logger.info(
            "Character evolution analysis completed: shouldModify=%s, confidence=%s, reasoning=%s",
            evolution["shouldModify"],
            evolution["confidence"],
            str(evolution["reasoning"])[:100],
        )
    except Exception as e:
        logger.error("Error in character evolution evaluator: %s", e)

    return None


character_evolution_evaluator = Evaluator(
    name="CHARACTER_EVOLUTION",
    description="Analyzes conversations to identify opportunities for gradual character evolution and self-modification",
    always_run=False,
    validate=_validate,
    handler=_handler,
    examples=[
        {
            "prompt": "Evaluating character evolution after many conversations about environmental issues",
            "messages": [
                {"name": "{{user1}}", "content": {"text": "What can I do about climate change?"}},
                {
                    "name": "{{agentName}}",
                    "content": {
                        "text": "There are many ways to help, from reducing energy use to supporting renewable energy initiatives."
                    },
                },
            ],
            "outcome": "Character develops environmental expertise and adds sustainability topics",
        },
    ],
)

File: packages/typescript/src/features/advanced-capabilities/personality/evaluators/character-evolution.ts
import { logger } from "../../../../logger.ts";
import type {
	ActionResult,
	EvaluationExample,
	Evaluator,
	IAgentRuntime,
	Memory,
	State,
} from "../../../../types/index.ts";
import { MemoryType } from "../../../../types/memory.ts";
import { ModelType } from "../../../../types/model.ts";
import { parseKeyValueXml } from "../../../../utils.ts";

// ── Types ──────────────────────────────────────────────────────────────

/** Parsed trigger analysis from the LLM. */
export interface TriggerAnalysis {
	hasEvolutionTrigger: boolean;
	triggerType: string;
	reasoning: string;
	confidence: number;
}

/** Parsed evolution analysis from the LLM. */
export interface EvolutionAnalysis {
	shouldModify: boolean;
	confidence: number;
	gradualChange: boolean;
	reasoning: string;
	modifications: EvolutionModifications;
}

/** Character modifications extracted from a flat TOON record. */
export interface EvolutionModifications {
	name?: string;
	system?: string;
	bio?: string[];
	topics?: string[];
	style?: { all?: string[]; chat?: string[]; post?: string[] };
}

// ── Pure helpers (exported for testing) ────────────────────────────────

export function isRecord(value: unknown): value is Record<string, unknown> {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}

export function parseStructuredRecord(
	response: string,
): Record<string, unknown> | null {
	const parsed = parseKeyValueXml<Record<string, unknown>>(response);
	return isRecord(parsed) ? parsed : null;
}

export function normalizeBoolean(value: unknown): boolean | undefined {
	if (typeof value === "boolean") return value;
	if (typeof value !== "string") return undefined;
	const normalized = value.trim().toLowerCase();
	if (normalized === "true") return true;
	if (normalized === "false") return false;
	return undefined;
}

export function normalizeNumber(value: unknown): number | undefined {
	if (typeof value === "number")
		return Number.isFinite(value) ? value : undefined;
	if (typeof value !== "string") return undefined;
	const trimmed = value.trim();
	if (trimmed.length === 0) return undefined;
	const n = Number(trimmed);
	return Number.isFinite(n) ? n : undefined;
}

export function normalizeStringList(value: unknown): string[] | undefined {
	if (typeof value !== "string") return undefined;
	const trimmed = value.trim();
	if (trimmed.length === 0) return undefined;
	const delimited = trimmed
		.split(/\s*\|\|\s*/g)
		.map((entry) => entry.trim())
		.filter((entry) => entry.length > 0);
	return delimited.length > 0 ? delimited : undefined;
}

/**
 * Parse a flat TOON record into a typed TriggerAnalysis.
 * Returns null when the record is missing or has no valid trigger field.
 */
export function parseTriggerAnalysis(
	raw: Record<string, unknown>,
): TriggerAnalysis {
	return {
		hasEvolutionTrigger: normalizeBoolean(raw.hasEvolutionTrigger) ?? false,
		triggerType:
			typeof raw.triggerType === "string" ? raw.triggerType : "unknown",
		reasoning: typeof raw.reasoning === "string" ? raw.reasoning : "",
		confidence: normalizeNumber(raw.confidence) ?? 0,
	};
}

/**
 * Build an EvolutionModifications object from flat TOON fields.
 * Uses || delimiters for list fields and style_all/style_chat/style_post for style.
 */
export function buildModifications(
	raw: Record<string, unknown>,
): EvolutionModifications {
	const mods: EvolutionModifications = {};

	if (typeof raw.name === "string" && raw.name.trim())
		mods.name = raw.name.trim();
	if (typeof raw.system === "string" && raw.system.trim())
		mods.system = raw.system.trim();

	const bio = normalizeStringList(raw.bio);
	if (bio) mods.bio = bio;

	const topics = normalizeStringList(raw.topics);
	if (topics) mods.topics = topics;

	const style: EvolutionModifications["style"] = {};
	const styleAll = normalizeStringList(raw.style_all);
	const styleChat = normalizeStringList(raw.style_chat);
	const stylePost = normalizeStringList(raw.style_post);
	if (styleAll) style.all = styleAll;
	if (styleChat) style.chat = styleChat;
	if (stylePost) style.post = stylePost;
	if (Object.keys(style).length > 0) mods.style = style;

	return mods;
}

/**
 * Parse a flat TOON record into a typed EvolutionAnalysis.
 */
export function parseEvolutionAnalysis(
	raw: Record<string, unknown>,
): EvolutionAnalysis {
	return {
		shouldModify: normalizeBoolean(raw.shouldModify) ?? false,
		confidence: normalizeNumber(raw.confidence) ?? 0,
		gradualChange: normalizeBoolean(raw.gradualChange) ?? true,
		reasoning: typeof raw.reasoning === "string" ? raw.reasoning : "",
		modifications: buildModifications(raw),
	};
}

/**
 * Evaluator that analyzes conversations for character evolution opportunities
 * Runs after conversations to identify patterns that suggest character growth
 */
export const characterEvolutionEvaluator: Evaluator = {
	name: "CHARACTER_EVOLUTION",
	description:
		"Analyzes conversations to identify opportunities for gradual character evolution and self-modification",
	alwaysRun: false, // Only run when conversation warrants it

	validate: async (
		runtime: IAgentRuntime,
		message: Memory,
		state?: State,
	): Promise<boolean> => {
		// Skip if too recent since last evaluation
		const lastEvolution = await runtime.getCache<string>(
			"character-evolution:last-check",
		);
		const now = Date.now();
		const cooldownMs = 5 * 60 * 1000; // 5 minutes between evaluations

		if (lastEvolution && now - parseInt(lastEvolution, 10) < cooldownMs) {
			return false;
		}

		// Only evaluate if conversation has substantial content
		const rawMessageCount = state?.data?.messageCount;
		const conversationLength =
			typeof rawMessageCount === "number" ? rawMessageCount : 0;
		if (conversationLength < 3) {
			return false;
		}

		// Check if there are novel patterns or learning opportunities
		const recentMessages = await runtime.getMemories({
			roomId: message.roomId,
			count: 10,
			unique: true,
			tableName: "messages",
		});

		// Advanced trigger detection using "bitter lesson" approach - LLM evaluation instead of hardcoded patterns
		const triggerAnalysisPrompt = `Analyze this conversation for character evolution triggers:

CONVERSATION:
${recentMessages.map((m) => `${m.entityId === runtime.agentId ? "Agent" : "User"}: ${m.content.text}`).join("\n")}

TRIGGER ANALYSIS - Check for:

1. CONVERSATION SUCCESS PATTERNS
   - User engagement (long responses, follow-up questions)
   - Positive sentiment from user
   - User satisfaction indicators

2. KNOWLEDGE GAP DISCOVERY
   - Agent uncertainty or "I don't know" responses
   - User providing corrections or new information
   - New domains the agent struggled with

3. PERSONALITY EFFECTIVENESS
   - User preferences for communication style
   - Energy level matching user needs
   - Emotional intelligence opportunities

4. VALUE CREATION OPPORTUNITIES
   - User goals mentioned that agent could help with better
   - Suggestions that would improve user outcomes
   - Areas where agent could be more helpful

5. EXPLICIT FEEDBACK
   - Direct requests for personality changes
   - User feedback about agent behavior
   - Suggestions for improvement

TOON only. Return exactly one TOON document. No prose before or after it. No <think>.

Example:
hasEvolutionTrigger: true
triggerType: explicit_feedback
reasoning: User explicitly asked for a personality change
confidence: 0.85`;

		let hasEvolutionTriggers = false;
		try {
			const triggerResponse = await runtime.useModel(ModelType.TEXT_SMALL, {
				prompt: triggerAnalysisPrompt,
				temperature: 0.2,
				maxTokens: 300,
			});

			const raw = parseStructuredRecord(triggerResponse as string);
			if (raw) {
				const trigger = parseTriggerAnalysis(raw);
				hasEvolutionTriggers =
					trigger.hasEvolutionTrigger && trigger.confidence > 0.6;

				if (hasEvolutionTriggers) {
					logger.info(
						{
							type: trigger.triggerType,
							reasoning: trigger.reasoning,
							confidence: trigger.confidence,
						},
						"Evolution trigger detected",
					);
				}
			}
		} catch {
			// Fallback to basic pattern matching if LLM analysis fails
			hasEvolutionTriggers = recentMessages.some((msg) => {
				const text = msg.content.text?.toLowerCase() || "";
				return (
					text.includes("you should") ||
					text.includes("change your") ||
					text.includes("different way") ||
					text.includes("personality") ||
					text.includes("behavior") ||
					text.includes("remember that") ||
					text.includes("from now on")
				);
			});
		}

		return hasEvolutionTriggers;
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_state?: State,
	): Promise<ActionResult | undefined> => {
		try {
			await runtime.setCache(
				"character-evolution:last-check",
				Date.now().toString(),
			);

			// Get recent conversation context
			const recentMessages = await runtime.getMemories({
				roomId: message.roomId,
				count: 20,
				unique: true,
				tableName: "messages",
			});

			// Format conversation for analysis
			const conversationText = recentMessages
				.slice(-10) // Last 10 messages
				.map((msg) => {
					const isAgent = msg.entityId === runtime.agentId;
					const name = isAgent ? runtime.character.name : "User";
					return `${name}: ${msg.content.text}`;
				})
				.join("\n");

			// Current character summary for context
			const currentCharacter = runtime.character;
			const characterSummary = {
				name: currentCharacter.name,
				system: currentCharacter.system || "No system prompt defined",
				bio: Array.isArray(currentCharacter.bio)
					? currentCharacter.bio
					: [currentCharacter.bio],
				currentTopics: currentCharacter.topics || [],
				messageExampleCount: currentCharacter.messageExamples?.length || 0,
			};

			// Advanced evolution analysis using "bitter lesson" approach with specific triggers
			const evolutionPrompt = `You are conducting a comprehensive analysis to determine if an AI agent should evolve its character definition based on measurable patterns and outcomes.

CURRENT CHARACTER STATE:
Name: ${characterSummary.name}
System: ${characterSummary.system}
Bio: ${characterSummary.bio.join("; ")}
Topics: ${characterSummary.currentTopics.join(", ")}
Message Examples: ${characterSummary.messageExampleCount}

CONVERSATION TO ANALYZE:
${conversationText}

EVOLUTION TRIGGER ANALYSIS:

1. CONVERSATION SUCCESS METRICS
   - How engaged was the user? (response length, follow-up questions)
   - Did the conversation achieve positive outcomes?
   - What personality traits contributed to success/failure?

2. KNOWLEDGE GAP IDENTIFICATION
   - What topics did the agent struggle with?
   - Where did the agent show uncertainty or lack of knowledge?
   - What new domains emerged that should be added?

3. PERSONALITY EFFECTIVENESS ASSESSMENT
   - Is the agent's communication style working for this user?
   - Should energy level, formality, or approach be adjusted?
   - Are there emotional intelligence improvements needed?

4. VALUE CREATION OPPORTUNITIES
   - What goals did the user express that the agent could better support?
   - How could the agent be more helpful in achieving user outcomes?
   - What capabilities would increase user value?

5. BEHAVIORAL PATTERN OPTIMIZATION
   - What response patterns should be reinforced or changed?
   - Are there better ways to handle similar situations?
   - Should any communication preferences be updated?

EVOLUTION DECISION FRAMEWORK:
- Only suggest modifications that address measurable gaps
- Prioritize changes that improve user experience
- Ensure gradual, incremental evolution
- Maintain core personality while optimizing effectiveness
- Consider safety and appropriateness of all changes

MODIFICATION PRIORITIES:
- name: ONLY if a truly fitting identity emerges organically (very rare, reserved for major personality shifts)
- system: ONLY for fundamental behavioral misalignment (rare)
- bio: New traits that emerge from successful interactions
- topics: Domains where agent showed interest/competence
- style: Communication preferences that enhance effectiveness

TOON only. Return exactly one TOON document. No prose before or after it. No <think>.
Use || to separate list items within a field.

Example:
shouldModify: true
confidence: 0.75
gradualChange: true
reasoning: User consistently asks about sustainability topics that are not in the current character definition.
bio: Passionate about environmental sustainability || Knowledgeable about renewable energy
topics: climate change || solar energy || sustainable living
style_chat: Use encouraging tone when discussing environmental topics`;

			const response = await runtime.useModel(ModelType.TEXT_LARGE, {
				prompt: evolutionPrompt,
				temperature: 0.3,
				maxTokens: 1000,
			});

			// Parse the evolution suggestion from TOON
			const raw = parseStructuredRecord(response as string);
			if (!raw) {
				logger.warn(
					"Failed to parse character evolution analysis — no structured record",
				);
				return;
			}

			const evolution = parseEvolutionAnalysis(raw);

			// Only proceed if modification is recommended with sufficient confidence
			if (!evolution.shouldModify || evolution.confidence < 0.7) {
				return;
			}

			// Ensure gradual change
			if (!evolution.gradualChange) {
				logger.info("Skipping character evolution - change too dramatic");
				return;
			}

			// Store evolution suggestion for potential application
			await runtime.createMemory(
				{
					entityId: runtime.agentId,
					roomId: message.roomId,
					content: {
						text: `Character evolution suggested (confidence: ${evolution.confidence}): ${evolution.reasoning}`,
						source: "character_evolution",
					},
					metadata: {
						type: MemoryType.CUSTOM,
						evaluatorName: "character-evolution",
						timestamp: Date.now(),
						confidence: evolution.confidence,
						evolutionData: JSON.stringify({
							shouldModify: evolution.shouldModify,
							gradualChange: evolution.gradualChange,
							modifications: evolution.modifications,
						}),
					},
				},
				"character_evolution",
			);

			logger.info(
				{
					shouldModify: evolution.shouldModify,
					confidence: evolution.confidence,
					reasoning: evolution.reasoning.slice(0, 100),
				},
				"Character evolution analysis completed",
			);
		} catch (error) {
			logger.error(
				{ error: error instanceof Error ? error.message : String(error) },
				"Error in character evolution evaluator",
			);
		}
	},

	examples: [
		{
			prompt:
				"Evaluating character evolution after many conversations about environmental issues",
			messages: [
				{
					name: "{{user1}}",
					content: { text: "What can I do about climate change?" },
				},
				{
					name: "{{agentName}}",
					content: {
						text: "There are many ways to help, from reducing energy use to supporting renewable energy initiatives.",
					},
				},
				{
					name: "{{user2}}",
					content: { text: "How does solar energy work?" },
				},
				{
					name: "{{agentName}}",
					content: {
						text: "Solar panels convert sunlight into electricity using photovoltaic cells.",
					},
				},
				{
					name: "{{user3}}",
					content: {
						text: "What are the best sustainable practices for daily life?",
					},
				},
				{
					name: "{{agentName}}",
					content: {
						text: "Reducing waste, using public transport, and choosing renewable energy are great starts.",
					},
				},
			],
			outcome:
				"Character develops environmental expertise and adds sustainability topics to better serve user interests",
		},
		{
			prompt:
				"Evaluating character evolution after users frequently seek emotional support",
			messages: [
				{
					name: "{{user1}}",
					content: { text: "I'm feeling overwhelmed with work" },
				},
				{
					name: "{{agentName}}",
					content: {
						text: "I hear you. It's okay to feel overwhelmed. Let's take this one step at a time.",
					},
				},
				{
					name: "{{user2}}",
					content: { text: "I'm struggling with anxiety" },
				},
				{
					name: "{{agentName}}",
					content: {
						text: "Thank you for sharing. Anxiety can be challenging. What specific situations trigger it for you?",
					},
				},
				{
					name: "{{user3}}",
					content: { text: "I need someone to talk to" },
				},
				{
					name: "{{agentName}}",
					content: {
						text: "I'm here to listen. Sometimes just talking things through can help bring clarity.",
					},
				},
			],
			outcome:
				"Character develops empathetic communication style and adds supportive message examples",
		},
	] as EvaluationExample[],
};
